File: src/hexa_kinematics/hexa_kinematics/joint_config.py
"""Per-joint servo configuration and default standing pose.

Loads the two YAMLs that live in ``hexa_description/config/``:

- ``geometry.yaml`` — under ``joints:``, per-joint-type servo center
  (URDF angle at the servo's physical zero) plus absolute lower / upper
  travel limits, all in intuitive per-joint degrees.
- ``standing_pose.yaml`` — per-joint-type default at-rest angle.

Both files express angles in **degrees**, in each joint's intuitive
sense. This module is the single source of truth for converting those
intuitive degrees into the IK-convention radians used by
``hexa_kinematics`` (see ``leg_geometry.py``). The same arithmetic is
inlined inside ``hexapod.urdf.xacro`` so the URDF stays a pure
mathematical presentation of the hexapod (joint zero = legs splayed
horizontally) without having to import any python.

Joint-type → IK-radian conversions:

- ``coxa``  — ``theta_coxa  =  radians(deg)``.
- ``femur`` — ``theta_femur = -radians(above_horizontal_deg)``; IK
  treats positive femur as tilting the foot toward ``-z``.
- ``tibia`` — ``theta_tibia =  pi - radians(interior_deg)``; matches
  the ``th_t = pi - gamma`` derivation in ``leg_ik.inverse_kinematics``.

Sign-aware swap: femur and tibia conversions are monotonically
decreasing, so an intuitive ``upper_limit_deg`` maps to a smaller
URDF-rad value than ``lower_limit_deg``. The loader reconciles this
with ``min/max`` after conversion.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from pathlib import Path

import yaml

from .leg_geometry import JointAngles


# Per-joint-type intuitive-center field name inside the YAML.
_CENTER_FIELD: dict[str, str] = {
    "coxa": "deg",
    "femur": "above_horizontal_deg",
    "tibia": "interior_deg",
}


@dataclass(frozen=True)
class JointLimits:
    """Servo configuration for one joint type, in IK-convention radians."""

    center: float    # rad — URDF angle at the servo's physical zero
    lower: float     # rad — URDF lower bound (always <= upper)
    upper: float     # rad — URDF upper bound
    effort: float    # Nm
    velocity: float  # rad/s


def _to_urdf_rad(joint_type: str, deg: float) -> float:
    """Convert an intuitive per-joint degree value to URDF-convention radians."""
    if joint_type == "coxa":
        return math.radians(deg)
    if joint_type == "femur":
        return -math.radians(deg)
    if joint_type == "tibia":
        return math.pi - math.radians(deg)
    raise ValueError(f"unknown joint type: {joint_type!r}")


def load_joint_limits(geometry_path: str | Path) -> dict[str, JointLimits]:
    """Parse ``geometry.yaml``'s ``joints:`` block into ``{joint_type: JointLimits}``.

    ``joint_type`` is one of ``"coxa"``, ``"femur"``, ``"tibia"``; the
    returned ``center``, ``lower``, and ``upper`` are in IK-convention
    radians, with ``lower <= center <= upper`` (the deg→rad sign flips
    on femur and tibia are absorbed by a ``min/max`` reconciliation).
    """
    with open(geometry_path) as f:
        raw = yaml.safe_load(f)
    joints = raw["joints"]
    out: dict[str, JointLimits] = {}
    for joint_type in ("coxa", "femur", "tibia"):
        cfg = joints[joint_type]
        center_deg = float(cfg[_CENTER_FIELD[joint_type]])
        lower_deg = float(cfg["lower_limit_deg"])
        upper_deg = float(cfg["upper_limit_deg"])

        center = _to_urdf_rad(joint_type, center_deg)
        a = _to_urdf_rad(joint_type, lower_deg)
        b = _to_urdf_rad(joint_type, upper_deg)
        lower, upper = (a, b) if a <= b else (b, a)

        if not (lower <= center <= upper):
            raise ValueError(
                f"{joint_type} servo center {center_deg:.2f}° lies outside "
                f"limit window [{lower_deg:.2f}°, {upper_deg:.2f}°]"
            )

        out[joint_type] = JointLimits(
            center=center,
            lower=lower,
            upper=upper,
            effort=float(cfg["effort"]),
            velocity=float(cfg["velocity"]),
        )
    return out


def load_standing_pose(
    standing_pose_path: str | Path,
    geometry_path: str | Path,
) -> JointAngles:
    """Parse ``standing_pose.yaml`` into ``(theta_coxa, theta_femur, theta_tibia)``.

    Angles are in IK-convention radians. Each joint's standing angle is
    validated against ``geometry.yaml``'s ``[lower, upper]`` window; a
    value outside that window raises ``ValueError`` so an inconsistent
    edit fails fast at startup instead of silently clipping inside the
    URDF.
    """
    with open(standing_pose_path) as f:
        raw = yaml.safe_load(f)
    limits = load_joint_limits(geometry_path)

    angles: dict[str, float] = {}
    for joint_type in ("coxa", "femur", "tibia"):
        cfg = raw[joint_type]
        theta = _to_urdf_rad(joint_type, float(cfg[_CENTER_FIELD[joint_type]]))
        lim = limits[joint_type]
        if not (lim.lower <= theta <= lim.upper):
            raise ValueError(
                f"standing pose {joint_type} angle {math.degrees(theta):.2f}° "
                f"lies outside servo range "
                f"[{math.degrees(lim.lower):.2f}°, {math.degrees(lim.upper):.2f}°]"
            )
        angles[joint_type] = theta

    return (angles["coxa"], angles["femur"], angles["tibia"])


def load_initial_pose(geometry_path: str | Path) -> dict[str, JointAngles]:
    """Parse ``geometry.yaml``'s ``initial_pose:`` block into per-leg ``JointAngles``.

    Returns ``{"l_front": (th_coxa, th_femur, th_tibia), ...}`` with one
    entry per leg, all in IK-convention radians. The YAML stores only
    the two reference coxa values (``l_front_deg`` and ``l_middle_deg``);
    the other four legs derive by the same mirror rules as the
    ``leg_joints_iface`` macro in ``hexapod.urdf.xacro`` and
    ``load_leg_specs``:

    - rear  : negate the reference coxa.deg (front/rear mirror)
    - r_*   : negate after the front/rear mirror (left/right mirror)

    Applying both leaves r_rear with the same sign as l_front. Femur
    and tibia are uniform across all six legs. Each per-leg angle is
    validated against ``geometry.yaml``'s ``[lower, upper]`` window so
    an inconsistent edit fails fast at startup.
    """
    with open(geometry_path) as f:
        raw = yaml.safe_load(f)
    limits = load_joint_limits(geometry_path)
    init = raw["initial_pose"]

    femur_theta = _to_urdf_rad("femur", float(init["femur"]["above_horizontal_deg"]))
    tibia_theta = _to_urdf_rad("tibia", float(init["tibia"]["interior_deg"]))
    for joint_type, theta in (("femur", femur_theta), ("tibia", tibia_theta)):
        lim = limits[joint_type]
        if not (lim.lower <= theta <= lim.upper):
            raise ValueError(
                f"initial pose {joint_type} angle {math.degrees(theta):.2f}° "
                f"lies outside servo range "
                f"[{math.degrees(lim.lower):.2f}°, {math.degrees(lim.upper):.2f}°]"
            )

    coxa_lim = limits["coxa"]
    out: dict[str, JointAngles] = {}
    for side in ("l", "r"):
        for name in ("front", "middle", "rear"):
            ref_deg = float(
                init["coxa"]["l_middle_deg" if name == "middle" else "l_front_deg"]
            )
            after_fr = -ref_deg if name == "rear" else ref_deg
            after_lr = -after_fr if side == "r" else after_fr
            coxa_theta = _to_urdf_rad("coxa", after_lr)
            if not (coxa_lim.lower <= coxa_theta <= coxa_lim.upper):
                raise ValueError(
                    f"initial pose coxa angle for {side}_{name} "
                    f"{math.degrees(coxa_theta):.2f}° lies outside servo range "
                    f"[{math.degrees(coxa_lim.lower):.2f}°, "
                    f"{math.degrees(coxa_lim.upper):.2f}°]"
                )
            out[f"{side}_{name}"] = (coxa_theta, femur_theta, tibia_theta)
    return out

File: src/hexa_kinematics_cpp/src/joint_config.cpp
#include "hexa_kinematics_cpp/joint_config.hpp"

#include <array>
#include <cmath>
#include <iomanip>
#include <sstream>
#include <stdexcept>

#include "hexa_kinematics_cpp/yaml_util.hpp"

namespace hexa_kinematics {

namespace {

// Per-joint-type intuitive-center field name inside the YAML.
const char* center_field(const std::string& joint_type) {
  if (joint_type == "coxa") return "deg";
  if (joint_type == "femur") return "above_horizontal_deg";
  if (joint_type == "tibia") return "interior_deg";
  throw std::runtime_error("unknown joint type: " + joint_type);
}

// Convert an intuitive per-joint degree value to URDF-convention radians.
double to_urdf_rad(const std::string& joint_type, double deg) {
  if (joint_type == "coxa") return deg * M_PI / 180.0;
  if (joint_type == "femur") return -deg * M_PI / 180.0;
  if (joint_type == "tibia") return M_PI - deg * M_PI / 180.0;
  throw std::runtime_error("unknown joint type: " + joint_type);
}

double degrees(double rad) { return rad * 180.0 / M_PI; }

std::string fmt_deg(double rad) {
  std::ostringstream s;
  s << std::fixed << std::setprecision(2) << degrees(rad);
  return s.str();
}

const std::array<std::string, 3> kJointTypes = {"coxa", "femur", "tibia"};

}  // namespace

std::map<std::string, JointLimits> load_joint_limits(
    const std::string& geometry_path) {
  const YAML::Node raw = load_file(geometry_path);
  const YAML::Node joints = raw["joints"];
  std::map<std::string, JointLimits> out;
  for (const auto& joint_type : kJointTypes) {
    const YAML::Node cfg = joints[joint_type];
    const std::string ctx = "joints." + joint_type;
    const double center_deg =
        require_scalar<double>(cfg, center_field(joint_type), ctx);
    const double lower_deg = require_scalar<double>(cfg, "lower_limit_deg", ctx);
    const double upper_deg = require_scalar<double>(cfg, "upper_limit_deg", ctx);

    const double center = to_urdf_rad(joint_type, center_deg);
    const double a = to_urdf_rad(joint_type, lower_deg);
    const double b = to_urdf_rad(joint_type, upper_deg);
    const double lower = std::min(a, b);
    const double upper = std::max(a, b);

    if (!(lower <= center && center <= upper)) {
      std::ostringstream msg;
      msg << std::fixed << std::setprecision(2) << joint_type << " servo center "
          << center_deg << "deg lies outside limit window [" << lower_deg
          << "deg, " << upper_deg << "deg]";
      throw std::runtime_error(msg.str());
    }

    JointLimits lim;
    lim.center = center;
    lim.lower = lower;
    lim.upper = upper;
    lim.effort = require_scalar<double>(cfg, "effort", ctx);
    lim.velocity = require_scalar<double>(cfg, "velocity", ctx);
    out[joint_type] = lim;
  }
  return out;
}

JointAngles load_standing_pose(const std::string& standing_pose_path,
                               const std::string& geometry_path) {
  const YAML::Node raw = load_file(standing_pose_path);
  const std::map<std::string, JointLimits> limits = load_joint_limits(geometry_path);

  std::array<double, 3> angles{};
  for (std::size_t i = 0; i < kJointTypes.size(); ++i) {
    const std::string& joint_type = kJointTypes[i];
    const YAML::Node cfg = raw[joint_type];
    const double theta = to_urdf_rad(
        joint_type, require_scalar<double>(cfg, center_field(joint_type),
                                           "standing_pose." + joint_type));
    const JointLimits& lim = limits.at(joint_type);
    if (!(lim.lower <= theta && theta <= lim.upper)) {
      throw std::runtime_error("standing pose " + joint_type + " angle " +
                               fmt_deg(theta) + "deg lies outside servo range [" +
                               fmt_deg(lim.lower) + "deg, " + fmt_deg(lim.upper) +
                               "deg]");
    }
    angles[i] = theta;
  }
  return {angles[0], angles[1], angles[2]};
}

std::map<std::string, JointAngles> load_initial_pose(
    const std::string& geometry_path) {
  const YAML::Node raw = load_file(geometry_path);
  const std::map<std::string, JointLimits> limits = load_joint_limits(geometry_path);
  const YAML::Node init = raw["initial_pose"];

  const double femur_theta = to_urdf_rad(
      "femur", require_scalar<double>(init["femur"], "above_horizontal_deg",
                                      "initial_pose.femur"));
  const double tibia_theta = to_urdf_rad(
      "tibia",
      require_scalar<double>(init["tibia"], "interior_deg", "initial_pose.tibia"));
  for (const auto& [joint_type, theta] :
       {std::pair<std::string, double>{"femur", femur_theta},
        std::pair<std::string, double>{"tibia", tibia_theta}}) {
    const JointLimits& lim = limits.at(joint_type);
    if (!(lim.lower <= theta && theta <= lim.upper)) {
      throw std::runtime_error("initial pose " + joint_type + " angle " +
                               fmt_deg(theta) + "deg lies outside servo range [" +
                               fmt_deg(lim.lower) + "deg, " + fmt_deg(lim.upper) +
                               "deg]");
    }
  }

  const JointLimits& coxa_lim = limits.at("coxa");
  const YAML::Node coxa_cfg = init["coxa"];
  std::map<std::string, JointAngles> out;
  for (const std::string side : {"l", "r"}) {
    for (const std::string name : {"front", "middle", "rear"}) {
      const double ref_deg = require_scalar<double>(
          coxa_cfg, (name == "middle") ? "l_middle_deg" : "l_front_deg",
          "initial_pose.coxa");
      const double after_fr = (name == "rear") ? -ref_deg : ref_deg;
      const double after_lr = (side == "r") ? -after_fr : after_fr;
      const double coxa_theta = to_urdf_rad("coxa", after_lr);
      if (!(coxa_lim.lower <= coxa_theta && coxa_theta <= coxa_lim.upper)) {
        throw std::runtime_error(
            "initial pose coxa angle for " + side + "_" + name + " " +
            fmt_deg(coxa_theta) + "deg lies outside servo range [" +
            fmt_deg(coxa_lim.lower) + "deg, " + fmt_deg(coxa_lim.upper) + "deg]");
      }
      out[side + "_" + name] = {coxa_theta, femur_theta, tibia_theta};
    }
  }
  return out;
}

}  // namespace hexa_kinematics
